### PR TITLE
openstack UPI: Automate Control Plane creation

### PR DESCRIPTION
**Pull request by pierreprinetti**
Create an Ansible playbook for deploying the Control Plane required for
a User Provided Infrastructure installation.

Implements https://jira.coreos.com/browse/OSASINFRA-787

### CODE REVIEW

**Comment by pierreprinetti** [comment]
/retest
/hold

while we figure out whether this can be merged during Feature Freeze.

File: upi/openstack/04_control-plane.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      - ip_address: "{{ os_subnet_range | next_nth_usable(6) }}"
+    with_indexed_items: "{{ cp_ports }}"
+
+  - name: 'Create the Control Plane servers'
+    os_server:
+      name: "{{ item.1 }}-{{ item.0 + 1 }}"
+      image: "{{ os_image_rhcos }}"
+      flavor: "{{ os_flavor_master }}"
+      auto_ip: no
+      userdata: "{{ item.1 }}-{{ item.0 + 1 }}-{{ cp_data }}"
```
Comment (tomassedovic) [inline upi/openstack/04_control-plane.yaml]: Same as in bootstrap, this needs the *contents* no the filename.

File: upi/openstack/inventory.yaml [hunk 2]
```diff
@@ -4,8 +4,14 @@ all:
...
       ansible_python_interpreter: "{{ansible_playbook_python}}"
 
+      # User-provided values
       os_subnet_range: '10.0.0.0/16'
       os_cluster_name: 'openshift'
+      os_flavor_master: 'm1.xlarge'
+      os_image_rhcos: 'rhcos'
+
+      # Number of provisioned Control Plane nodes
+      os_cp_nodes_number: 2
```
Comment (tomassedovic) [inline upi/openstack/inventory.yaml]: Oh sorry, I should have noticed this sooner. Control plane (masters) must have a minimum of three nodes.

File: upi/openstack/04_control-plane.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+#
+# ansible
+# openstacksdk
+# netaddr
+
+- hosts: all
+
+  vars:
+    # The provided filename will be concatenated with the Control Plane node
+    # name and its 1-indexed serial number.
```
Comment (tomassedovic) [inline upi/openstack/04_control-plane.yaml]: Please make it 0-indexed. That's what IPI as well as the UPI doc do currently.

File: upi/openstack/04_control-plane.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      - ip_address: "{{ os_subnet_range | next_nth_usable(6) }}"
+    with_indexed_items: "{{ cp_ports }}"
+
+  - name: 'Create the Control Plane servers'
+    os_server:
+      name: "{{ item.1 }}-{{ item.0 + 1 }}"
+      image: "{{ os_image_rhcos }}"
+      flavor: "{{ os_flavor_master }}"
+      auto_ip: no
+      userdata: "{{ lookup('file', {{ item.1 }}-{{ item.0 + 1 }}-{{ cp_data }}) | string }}"
```
Comment (tomassedovic) [inline upi/openstack/04_control-plane.yaml]:
```
fatal: [localhost]: FAILED! => {"msg": "template error while templating string: expected token ':', got '}'. String: {{ lookup('file', {{ item.1 }}-{{ item.0 + 1 }}-{{ cp_data }}) | string }}"}
```

This seems to work for me:

```
      userdata: "{{ lookup('file', [item.1, item.0, cp_data] | join('-')) | string }}"
```

File: upi/openstack/04_control-plane.yaml (hunk 3 above)
Comment (pierreprinetti) [reply upi/openstack/04_control-plane.yaml]: true, thanks. I was taking IPI as an example instead.

**Comment by pierreprinetti** [comment]
@tomassedovic 
I have addressed your comments. thank you very much for your review. Please take a look!

**Pull request merged**